daemon/src/close_position/protocol.rs

```diff
@@ -0,0 +1,226 @@
+use model::CollaborativeSettlement;
+use model::OrderId;
+use model::Price;
+use serde::Deserialize;
+use serde::Serialize;
+use xtra::Address;
+use xtra_libp2p::Endpoint;
+use xtra_libp2p::OpenSubstream;
+
+pub async fn dialer(
+    endpoint: Address<Endpoint>,
+    order_id: OrderId,
+    counterparty: PeerId,
+    close_position_tx: ClosePositionTransaction,
+) -> Result<CollaborativeSettlement, DialerFailed> {
+    let substream = endpoint
+        .send(OpenSubstream::single_protocol(counterparty, PROTOCOL))
+        .await
+        .context("Endpoint is disconnected")?
+        .context("Failed to open substream")?;
+    let mut framed = asynchronous_codec::Framed::new(
+        substream,
+        asynchronous_codec::JsonCodec::<DialerMessage, ListenerMessage>::new(),
+    );
+
+    let unsigned_tx = close_position_tx.unsigned_transaction();
+
+    framed
+        .send(DialerMessage::Propose(Propose {
+            id: order_id,
+            price: close_position_tx.price(),
+            unsigned_tx: unsigned_tx.clone(),
+        }))
+        .await
+        .context("Failed to send Msg0")?;
+
+    // TODO: We will need to apply a timeout to these. Perhaps we can put a timeout generally into
+    // "reading from the substream"?
+    if let Decision::Reject = framed
+        .next()
+        .await
+        .context("End of stream while receiving Msg1")?
+        .context("Failed to decode Msg1")?
+        .into_msg1()?
+    {
+        return Err(DialerFailed::Rejected);
+    }
+
+    framed
+        .send(DialerMessage::Msg2(Msg2 {
+            dialer_signature: close_position_tx.own_signature(),
+        }))
+        .await
+        .context("Failed to send Msg2")?;
+
+    let msg3 = match framed.next().await {
+        Some(Ok(msg)) => msg.into_msg3()?,
+        Some(Err(_)) | None => {
+            return Err(DialerFailed::AfterSendingSignature {
+                unsigned_tx,
+                error: anyhow!("failed to received msg3"),
+            });
+        }
+    };
+
+    let close_position_tx =
+        match close_position_tx.recv_counterparty_signature(msg3.listener_signature) {
+            Ok(close_position_tx) => close_position_tx,
+            Err(error) => {
+                return Err(DialerFailed::AfterSendingSignature { unsigned_tx, error });
+            }
+        };
+
+    let settlement =
+        close_position_tx
+            .finalize()
+            .map_err(|e| DialerFailed::AfterSendingSignature {
+                unsigned_tx,
+                error: e,
+            })?;
+
+    Ok(settlement)
+}
+
+#[derive(Debug)]
+pub enum DialerFailed {
+    Rejected,
+    AfterSendingSignature {
+        unsigned_tx: Transaction,
+        error: anyhow::Error,
+    },
+    BeforeSendingSignature {
+        source: anyhow::Error,
+    },
+}
+
+impl From<anyhow::Error> for DialerFailed {
+    fn from(source: anyhow::Error) -> Self {
+        Self::BeforeSendingSignature { source }
+    }
+}
```

Comment:
🔧 Similarly, very unintuitive to me.

Reply:
again, failure before signature is the typical error path, the other one is an edge case